docs/architecture/adr-016-validator-consensus-key-rotation.md

```diff
@@ -0,0 +1,122 @@
+
+## Decision
+
+### Pseudo procedure for consensus key rotation
+
+- create new random consensus key.
+- create and broadcast a transaction with a `MsgRotateConsPubKey` that states the new consensus key is now coupled with the validator operator with signature from the validator's operator key.
+- old consensus key becomes unable to participate on consensus immediately after the update of key mapping state on-chain.
+- start validating with new consensus key.
+- validators using HSM and KMS should update the consensus key in HSM to use the new rotated key after the height `h` when `MsgRotateConsPubKey` committed to the blockchain.
+
+
+### Considerations
+
+- consensus key mapping information management strategy
+    - store history of each key mapping changes in the kvstore.
+    - the state machine can search corresponding consensus key paired with given validator operator for any arbitrary height in a recent unbonding period.
+    - the state machine does not need any historical mapping information which is past more than unbonding period.
+- limits
+    - a validator cannot rotate its consensus key more than `MaxConsPubKeyRotations` time for any unbonding period, to prevent spam.
+    - parameters can be decided by governance and stored in genesis file.
+- key rotation fee
+    - a validator should pay `KeyRotationFee` to rotate the consensus key which is calculated as below
+    - `KeyRotationFee` = `InitialKeyRotationFee` * 2^(number of rotations in `ConsPubKeyRotationHistory` in recent unbonding period)
+- evidence module
+    - evidence module can search corresponding consensus key for any height from slashing keeper so that it can decide which consensus key is supposed to be used for given height.
+- abci.ValidatorUpdate
+    - tendermint already has ability to change a consensus key by ABCI communication(`ValidatorUpdate`).
+    - validator consensus key update can be done via creating new + delete old by change the power to zero.
+    - therefore, we expect we even do not need to change tendermint codebase at all to implement this feature.
+- new genesis parameters in `staking` module
+    - `MaxConsPubKeyRotations` : maximum number of rotation can be executed by a validator in recent unbonding period. default value 10 is suggested(11th key rotation will be rejected)
+    - `InitialKeyRotationFee` : the initial key rotation fee when no key rotation has happened in recent unbonding period. default value 1atom is suggested(1atom fee for the first key rotation in recent unbonding period)
+
+
+### Workflow
+
+1. The validator generates a new consensus keypair.
+2. The validator generates and signs a `MsgRotateConsPubKey` tx with their operator key and new ConsPubKey
+
+    ```go
+    type MsgRotateConsPubKey struct {
```

Comment:
Who is the signer of this message?

Reply:
The validator account key is the signer of `MsgRotateConsPubKey`